src/MongoDB.Driver/Linq/FieldExpressionFlattener.cs

```diff
@@ -86,5 +80,43 @@ protected internal override Expression VisitField(FieldExpression node)
 
             return node;
         }
+
+        // private methods
+        private static string FormatArrayIndex(object index)
+        {
+            if (index is int intIndex)
+            {
+                return GetFormattedIndexOrThrow(intIndex);
+            }
+
+            if (index is long longIndex)
+            {
+                return GetFormattedIndexOrThrow(longIndex);
+            }
+
+            throw new NotSupportedException("Array indexes must be int or long.");
+        }
+
+        private static string GetFormattedIndexOrThrow(long index) 
+        {
+            switch (index)
+            {
+                // We've treated -1 as meaning $ operator. We can't break this now,
+                // so, specifically when we are flattening fields names, this is 
+                // how we'll continue to treat -1.
+                case var _ when index < -1L:
+                {
+                    throw new NotSupportedException("Array indexes must be more or equal than -1.");
+                }
+                case var _ when index == -1L:
+                {
+                    return "$";
+                }
+                default:
+                {
+                    return index.ToString(NumberFormatInfo.InvariantInfo);
+                }
+            }
```

Comment:
This might be an abuse of the the switch statement. See if you think the following is more readable:

    if (index < 0)
    {
        if (index == -1)
        {
            // We've treated -1 as meaning $ operator. We can't break this now,
            // so, specifically when we are flattening fields names, this is 
            // how we'll continue to treat -1.
            return "$";
        }
        else
        {
            throw new NotSupportedException("Array indexes cannot be negative.");
        }
    }

    return index.ToString(NumberFormatInfo.InvariantInfo);

Just a suggestion. I will approve either way.

Reply:
My idea was that having conditions on the same nesting level easier to understand than otherwise. So, for me, my way is more readable, but the difference obviously is very small

Reply:
Sure... there are lots of ways to write code correctly.

My perspective was to handle the unusual cases first, and emphasize that calling ToString was the normal case.

No need to change though. I'm fine with your alternative also.